firebase-firestore/src/androidTest/java/com/google/firebase/firestore/SnapshotListenerSourceTest.java

```diff
@@ -0,0 +1,654 @@
+    snapshot = cacheAccumulator.await();
+    assertEquals(expected, querySnapshotToValues(snapshot));
+
+    defaultAccumulator.assertNoAdditionalEvents();
+    cacheAccumulator.assertNoAdditionalEvents();
+
+    defaultRegistration.remove();
+    cacheRegistration.remove();
+  }
+
+  @Test
+  public void canListenToCompositeIndexQueriesFromCache() {
+    CollectionReference collection =
+        testCollectionWithDocs(map("a", map("k", "a", "sort", 0), "b", map("k", "b", "sort", 1)));
+    Query query = collection.whereLessThanOrEqualTo("k", "a").whereGreaterThanOrEqualTo("sort", 0);
+
+    waitFor(collection.get()); // Populate the cache.
+
+    EventAccumulator<QuerySnapshot> accumulator = new EventAccumulator<>();
+    SnapshotListenOptions options = optionSourceFromCache();
+    ListenerRegistration listener = query.addSnapshotListener(options, accumulator.listener());
+
+    QuerySnapshot snapshot = accumulator.await();
+    assertEquals(asList(map("k", "a", "sort", 0L)), querySnapshotToValues(snapshot));
+    assertTrue(snapshot.getMetadata().isFromCache());
+
+    accumulator.assertNoAdditionalEvents();
+    listener.remove();
+  }
+
+  @Test
+  public void canRaiseInitialSnapshotFromCacheEvenIfItIsEmpty() {
+    CollectionReference collection = testCollection();
+
+    QuerySnapshot snapshot = waitFor(collection.get()); // Populate the cache
+    assertEquals(asList(), querySnapshotToValues(snapshot)); // Precondition check.
+
+    EventAccumulator<QuerySnapshot> accumulator = new EventAccumulator<>();
+    SnapshotListenOptions options = optionSourceFromCache();
+    ListenerRegistration listener = collection.addSnapshotListener(options, accumulator.listener());
+
+    snapshot = accumulator.await();
+    assertEquals(asList(), querySnapshotToValues(snapshot));
+    assertTrue(snapshot.getMetadata().isFromCache());
+
+    accumulator.assertNoAdditionalEvents();
+    listener.remove();
+  }
+
+  @Test
+  public void willNotBeTriggeredByTransactionsWhileListeningToCache() {
+    CollectionReference collection = testCollection();
+
+    EventAccumulator<QuerySnapshot> accumulator = new EventAccumulator<>();
+    SnapshotListenOptions options = optionSourceFromCacheAndIncludeMetadataChanges();
+    ListenerRegistration listener = collection.addSnapshotListener(options, accumulator.listener());
+
+    QuerySnapshot snapshot = accumulator.await();
+    assertEquals(asList(), querySnapshotToValues(snapshot));
+
+    DocumentReference docRef = collection.document();
+    // Use a transaction to perform a write without triggering any local events.
+    docRef
+        .getFirestore()
+        .runTransaction(
+            transaction -> {
+              transaction.set(docRef, map("k", "a"));
+              return null;
+            });
+
+    // There should be no events raised
+    accumulator.assertNoAdditionalEvents();
+    listener.remove();
+  }
+
+  @Test
+  public void shareServerSideUpdatesWhenListeningToBothCacheAndDefault() {
```

Comment:
You should also **unlisten** DEFAULT, and ensure CACHE does not receive updates.

And also the inverse.

Reply:
unlistening to one while not affecting another one is already tested above: "canUnlistenToDefaultSourceWhileStillListeningToCache" and "canUnlistenToCachetSourceWhileStillListeningToServer" which inlcuses adding doc after unlistening to one of the listeners.

Reply:
Sorry, I didn't mention this was with changes being done through transaction. It is true that CACHE should listen for local document updates, but it should not if done through transaction (appears like a remote update). So what I am looking for is whether unlistening a DEFAULT will continue to receive **remote** updates on CACHE listener. And vica versa.

I think the test you are referring to only captures updates done locally.

Reply:
Aha. I saw your point. Basically, we are checking after the default listener is removed, the connection should be closed, and server side updates (mocking by another transaction) should not trigger the cache listener. 
I will add these test cases.

Reply:
I am having a second sound. 
"I think the test you are referring to only captures updates done locally." -> Actually it is not. transactions doesn't kick local cache, and no snapshots are raised when it is executed (tested in the previous case). both default and cache raised the snapshot when we receive updates from the server, and the "fromCache" is false.

as for "unlistening a DEFAULT will continue to receive remote updates on CACHE listener", this should not happen. Once DEFAULT is unlisted, connection is closed, no remote updates should be received.  this is tested in "canUnlistenToDefaultSourceWhileStillListeningToCache".

Reply:
Ok. No change required.

From offline conversation I understand that there is no easy way to verify absence of event, and that because another test verifies snapshot being raised from cache and remote (via metadata), we area actualy already testing this.

I am fine with this.